Spigot-Server-Patches/0594-Damage-with-any-cause.patch

```diff
@@ -0,0 +1,528 @@
++    }
++
++    // Paper end
++
+     @Override
+     public Location getEyeLocation() {
+         Location loc = getLocation();
+diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+index 0f0d70c0622558740086f0f7cf89e5500128f955..4d584c423835e2e2d8b038676904bfed30571671 100644
+--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
++++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+@@ -99,6 +99,7 @@ import org.bukkit.craftbukkit.potion.CraftPotionUtil;
+ import org.bukkit.craftbukkit.util.CraftDamageSource;
+ import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+ import org.bukkit.craftbukkit.util.CraftNamespacedKey;
++import org.bukkit.craftbukkit.util.DamageUtils;
+ import org.bukkit.entity.AbstractHorse;
+ import org.bukkit.entity.Animals;
+ import org.bukkit.entity.AreaEffectCloud;
+@@ -886,10 +887,16 @@ public class CraftEventFactory {
+     }
+ 
+     private static EntityDamageEvent handleEntityDamageEvent(Entity entity, DamageSource source, Map<DamageModifier, Double> modifiers, Map<DamageModifier, Function<? super Double, Double>> modifierFunctions, boolean cancelled) {
++        // Paper start
++        Entity damager = entityDamage;
++        entityDamage = null;
++        // Paper end
+         if (source.isExplosion()) {
+             DamageCause damageCause;
+-            Entity damager = entityDamage;
+-            entityDamage = null;
++            // Paper start - moved above conditional block
++            //Entity damager = entityDamage;
++            //entityDamage = null;
++            // Paper end
+             EntityDamageEvent event;
+             if (damager == null) {
+                 event = new EntityDamageByBlockEvent(null, entity.getBukkitEntity(), DamageCause.BLOCK_EXPLOSION, modifiers, modifierFunctions);
+@@ -912,7 +919,12 @@ public class CraftEventFactory {
+             }
+             return event;
+         } else if (source instanceof EntityDamageSource) {
+-            Entity damager = source.getEntity();
++
++            // Paper start
++            //Entity damager = source.getEntity();
++            damager = source.getEntity();
++            // Paper end
++
+             DamageCause cause = (source.isSweep()) ? DamageCause.ENTITY_SWEEP_ATTACK : DamageCause.ENTITY_ATTACK;
+ 
+             if (source instanceof EntityDamageSourceIndirect) {
+@@ -928,7 +940,17 @@ public class CraftEventFactory {
+ 
+             return callEntityDamageEvent(damager, entity, cause, modifiers, modifierFunctions, cancelled);
+         } else if (source == DamageSource.OUT_OF_WORLD) {
+-            EntityDamageEvent event = new EntityDamageByBlockEvent(null, entity.getBukkitEntity(), DamageCause.VOID, modifiers, modifierFunctions);
++
++            // Paper start - OUT_OF_WORLD now may have entity damager source
++            EntityDamageEvent event;
++            if (damager == null) {
```

Comment:
Ternary operator?

Reply:
No, the constructor call is quite long 100+ characters. no real benefit of making it oneliner

Reply:
Except maintainability and the minimal diff policy...